pkg/controllers/sync/controller.go

```diff
@@ -1104,58 +1111,120 @@ func (s *SyncController) reconcileClusterForCascadingDeletion(
 		return worker.StatusAllOK
 	}
 
-	// cascading-delete is enabled, wait for member objects to be deleted
+	if !clusterutil.IsClusterReady(&cluster.Status) && cascadingdeletion.IsCascadingDeleteEnabled(cluster) {
+		return worker.Result{RequeueAfter: &s.cascadingDeletionRecheckDelay}
+	}
+
+	// cascading-delete is enabled, delete the member objects managed by admiral
 	ftcLister := s.ftcManager.GetFederatedTypeConfigLister()
 	ftcs, err := ftcLister.List(labels.Everything())
 	if err != nil {
-		logger.Error(err, "failed to get ftc lister")
+		logger.Error(err, "Failed to get ftc lister")
 		return worker.StatusError
 	}
 
-	remainingByGVK := make(map[string]string, len(ftcs))
-	for _, ftc := range ftcs {
-		gvk := ftc.GetSourceTypeGVK().String()
-		resourceLister, hasSynced, exists := s.fedInformerManager.GetResourceLister(
-			ftc.GetSourceTypeGVK(),
-			cluster.Name,
-		)
-		if !exists {
-			remainingByGVK[gvk] = fmt.Sprintf("failed to get resource lister for %s", gvk)
-			continue
-		}
+	// concurrent operations for each ftc
+	wg := sync.WaitGroup{}
+	syncMap := sync.Map{}
+	for i := range ftcs {
+		wg.Add(1)
+		go func(ftc *fedcorev1a1.FederatedTypeConfig) {
```

Comment:
Shall we extract this into a separate function?

Reply:
This should be considered the main logic, but it is encapsulated in the style of an anonymous function for concurrency. I think it does not need to be extracted separately.